test/functional/p2p_post_invalidate_sync.py

```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#
+# Test invalidateblock
+#
+
+from test_framework.test_framework import BitcoinTestFramework
+
+class InvalidateBlockTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+        self.extra_args = [['-debug', '-whitelist=127.0.0.1'],['-debug', '-whitelist=127.0.0.1']]
+
+    def run_test(self):
+        self.nodes[0].generate(1) # Leave IBD
+        self.sync_all()
+
+        cnt = self.nodes[0].getblockcount()
+
+        node1blocks = self.nodes[1].generate(18)
+
+        self.sync_all()
+        if (self.nodes[0].getblockcount() != cnt + 18):
+            raise AssertionError("Failed to sync initial blocks")
+
+        self.nodes[0].invalidateblock(node1blocks[0])
+        self.nodes[1].invalidateblock(node1blocks[0])
+
+        if (self.nodes[0].getblockcount() != cnt):
+            raise AssertionError("Failed to invalidate initial blocks")
+
+        # The test framework is a little dense wrt mining on nodes without wallets -
```

Comment:
This comment is wrong. All nodes will generate to a deterministic address (whether wallet is enabled or not).

Other ways to achieve this:
- call `self.nodes[0].generatetoaddress(17, self.nodes[0].getnewaddress()) to get a new address (and hence new block hash)
- call `sleep(1)` to get a new timestamp (and hence new block hash)
- ~use `setmocktime()` to get a new timestamp (and hence new block hash)~ probably not a good idea if you're testing p2p behaviour.

Reply:
also: I'd gently encourage you not to call other contributors' code 'a little dense'

Reply:
I tried getnewaddress first - sadly this is just an issue for non-wallet nodes. Wallet'd nodes will generate a *new* deterministic address, but the test framework just hard-codes a bunch. That's totally fine, but it is a bit dense. I'm also a bit dense, and so is lots of code I write, cause its just a hack to get it to work. Didn't mean it at all as an insult, just describing the state of it.

Reply:
It's not dense. It's how we get tests to work with wallet enabled and disabled.

Can you remove the incorrect comment please?

Reply:
Fair, likely should not have used the word "dense". I cleared up the comment, though I still think this approach is the easiest?